File: Blockchain/Chaincode/hospital.go
package main

import (
	"encoding/json"
	"fmt"
	. "fmt"
	"time"
)

func (c *Chaincode) CreateNewReport(ctx CustomTransactionContextInterface, patientID, refDoctor string) (string, error) {
	if ctx.GetData() == nil {
		return "", Errorf("Patient of ID %v doesn't exists", patientID)
	}
	id := REPORT + getSafeRandomString(ctx.GetStub())
	report := Report{
		DocTyp:      REPORT,
		ID:          id,
		PatientID:   patientID,
		Status:      "0",
		RefDoctorID: refDoctor,
		Comments:    make(map[string]string),
		CreateTime:  time.Now().Unix(),
		UpdateTime:  time.Now().Unix(),
	}
	reportAsByte, _ := json.Marshal(report)
	return report.ID, ctx.GetStub().PutState(id, reportAsByte)
}

func (c *Chaincode) StartTreatment(ctx CustomTransactionContextInterface, treatmentID, supervisor string) error {
	if ctx.GetData() == nil {
		return Errorf("Treatment with ID %v doesn't exists", treatmentID)
	}
	var treatment Treatment
	json.Unmarshal(ctx.GetData(), &treatment)
	if treatment.Status != 0 {
		return Errorf("Cannot start allready started treatment")
	}
	treatment.Supervisor = supervisor
	treatment.Status = 1
	treatment.UpdateTime = time.Now().Unix()

	treatmentAsByte, _ := json.Marshal(treatment)

	return ctx.GetStub().PutState(treatment.ID, treatmentAsByte)
}

// Create a new hospital
func (c *Chaincode) CreateHospital(ctx CustomTransactionContextInterface, id string, name string, email string, licenseNo string, phoneNumber string, address string) error {
	// Check if the hospital with the given ID already exists
	exists, err := c.hospitalExists(ctx, id)
	if err != nil {
		return err
	}
	if exists {
		return fmt.Errorf("the hospital with ID %s already exists", id)
	}

	// Create a new hospital struct
	hospital := Hospital{
		ID:          id,
		Name:        name,
		Email:       email,
		LicenseNo:   licenseNo,
		PhoneNumber: phoneNumber,
		Address:     address,
		CreatedAt:   time.Now().Unix(),
		UpdatedAt:   time.Now().Unix(),
	}

	// Convert the hospital struct to JSON format
	hospitalJSON, err := json.Marshal(hospital)
	if err != nil {
		return err
	}

	// Save the hospital to the world state
	err = ctx.GetStub().PutState(id, hospitalJSON)
	if err != nil {
		return err
	}

	return nil
}

// hospitalExists returns true if the hospital with given ID exists in the ledger
func (c *Chaincode) hospitalExists(ctx CustomTransactionContextInterface, id string) (bool, error) {
	hospitalJSON, err := ctx.GetStub().GetState(id)
	if err != nil {
		return false, err
	}
	return hospitalJSON != nil, nil
}

// getAllHospitals returns all hospitals in the ledger
func (s *Chaincode) GetAllHospitals(ctx CustomTransactionContextInterface) ([]*Hospital, error) {
	hospitalIterator, err := ctx.GetStub().GetStateByPartialCompositeKey(HOSPITAL, []string{})
	if err != nil {
		return nil, fmt.Errorf("failed to get all hospitals: %v", err)
	}
	defer hospitalIterator.Close()

	var hospitals []*Hospital
	for hospitalIterator.HasNext() {
		hospitalResult, err := hospitalIterator.Next()
		if err != nil {
			return nil, fmt.Errorf("failed to iterate over all hospitals: %v", err)
		}

		hospital := new(Hospital)
		err = json.Unmarshal(hospitalResult.GetValue(), hospital)
		if err != nil {
			return nil, fmt.Errorf("failed to unmarshal hospital data: %v", err)
		}

		hospitals = append(hospitals, hospital)
	}

	return hospitals, nil
}

File: Blockchain/node-api/api/routes/hospital.js
const express = require("express");
const { v4: uuidv4 } = require("uuid");
const isAuthorize = require("../util/isAuthorizedAdmin");
const { contract } = require("../contract");

const routes = express.Router();

routes.post("/createreport", (req, res) => {
  contract(
    req.body.clientId,
    "INVOKE",
    ["CreateNewReport", req.body.patient_id, req.body.ref_doctor],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        res.status(200).json(payload.toString());
      }
    }
  );
});

routes.put("/starttreatment", (req, res) => {
  contract(
    req.body.clientId,
    "INVOKE",
    ["StartTreatment", req.headers.treatment_id, req.body.supervisor],
    (err, payload) => {
      if (err) {
        res.status(500).json(err);
      } else {
        res.status(200).json({
          message: `successfully started the treatment`,
        });
      }
    }
  );
});

// createHospital route
routes.post("/create-hospital", async (req, res) => {
  const isAuthorizedToAccess = await isAuthorize(req.headers.authorization);
  if (isAuthorizedToAccess) {
    contract(
      req.body.clientId,
      "INVOKE",
      [
        "CreateHospital",
        uuidv4(),
        req.body.name,
        req.body.email,
        req.body.licenseNo,
        req.body.phoneNumber,
        req.body.address,
      ],
      (err, payload) => {
        if (err) {
          console.log(err);
          res.status(500).json(err);
        } else {
          res.status(200).json({
            message: `Successfully created hospital ${req.body.name}`,
            data: payload,
          });
        }
      }
    );
  } else {
    res.status(500).json({ message: "unauthorized!" });
  }
});

// getAllHospitals route
routes.get("/hospitals", (req, res) => {
  contract(req.query.clientId, "QUERY", ["GetAllHospitals"], (err, payload) => {
    if (err) {
      res.status(500).json(err);
    } else {
      const hospitals = JSON.parse(payload);
      res.status(200).json(hospitals);
    }
  });
});

module.exports = routes;
